### ISSUE TITLE
Google Play action required - background location access

### ISSUE THREAD

**Issue by misaochan** [commons-app/apps-android-commons#5031]
@nicolas-raoul and I received, over the weekend, more emails from the Play Store regarding non-compliance. Snippet below:

> Issue found: Feature does not meet requirements for background location access
> Based on our review, your declared feature does not meet the requirements for background location access.
> 
> Please remove the background location permission requested and submit an update to your app. When declaring a feature for background location access, please note the following:
> 
> Your selected feature should deliver clear value to the user and be important to the core functionality, or main purpose of the app. Without this core feature, the app is "broken" or rendered unusable.
> You should also consider if users would expect the app to access their location in the background, and if you can deliver the same experience without accessing location in the background.

Oddly, both of us also looked at the manifest and could not find any mention of ACCESS_BACKGROUND_LOCATION in the manifest. Does anyone have any ideas for what the cause could be? @neslihanturan @madhurgupta10 @whym

**Comment by madhurgupta10** [comment]
@misaochan I think the permission is added automatically by Mapbox SDK

**Comment by misaochan** [comment]
Hmm, okay. By the way, there is actually another part of the email:

> Issue found: Missing information in prominent disclosure
> Your prominent disclosure must appear before your app’s location runtime permission, and should tell the user which feature(s) will use location in the background. Based on our review, your app’s prominent disclosure did not meet this criteria.
> 
> Based on our review, your prominent disclosure doesn’t comply with our policy requirements.The language in the disclosure MUST include the following elements:
> 
> The term “location”
> Indication that the nature of usage is in the background by using one of the following phrases “background” / “when the app is closed” / “always in use” / “when the app is not in use”
> A list of all the features that use location in the background
> If you extend permitted usage to ads, you must include the following: “used to provide ads / support advertising / support ads.” (Choose the most accurate phrasing)
> To meet the policy requirements, it’s recommended that you reference one of the following example formats, the second example includes the use of location for ads (choose the most relevant phrasing):
> 
> “[This app] collects location data to enable [“feature“], [“feature“], & [“feature“] even when the app is closed or not in use.”
> “[This app] collects location data to enable [“feature“], [“feature“], & [“feature“] even when the app is closed or not in use and it is also used to support advertising.”
> Example: “Fitness Funds collects location data to enable fitness tracking even when the app is closed or not in use.”

![IN_APP_EXPERIENCE-1504](https://user-images.githubusercontent.com/3611199/183360542-1d82b470-ccc0-4f75-ae6b-8eef665df009.png)

However, the way I read it, that was superseded by the "Feature does not meet requirements for background location access" issue - i.e. even if we fix the disclosure, we are still considered to not be eligible for background location access.

I guess a few different things that we can try are:
- Fixing the disclosure anyway, and explaining that Mapbox requires it? This should be fairly simple, however I'm not sure that it will solve the issue
- Creating an issue on Mapbox's GitHub to see if anyone knows a way around it (takes longer)
- Appealing to Google Play and explaining that our Nearby feature in fact "deliver[s] clear value to the user and be important to the core functionality, or main purpose of the app"

**Comment by nicolas-raoul** [comment]
Maybe use `<uses-permission android:name="android.permission.ACCESS_BACKGROUND_LOCATION" tools:node="remove" />` ? Note the `remove` part.
Use case from [the documentation](https://developer.android.com/studio/build/manage-manifests): "_necessary when you discover an element in your merged manifest that you don't need, and it was provided by a lower-priority manifest file that's out of your control (such as an imported library)._"

Also, maybe we should do this every time we use location?
https://developer.android.com/about/versions/oreo/background-location-limits#tuning-behavior

**Comment by nicolas-raoul** [comment]
By any chance, does [LocationServiceManager](https://github.com/commons-app/apps-android-commons/blob/a0ff15cdc23e32763139dfdf86e2f9e1b9b913e1/app/src/main/java/fr/free/nrw/commons/location/LocationServiceManager.java) or any similar class somehow survives (and maybe even run) when the app is not visible on screen?

**Comment by misaochan** [comment]
I don't think it's possible for us to implement background location checking without it being possible in the manifest, right? I wonder if perhaps the best next step is to try `<uses-permission android:name="android.permission.ACCESS_BACKGROUND_LOCATION" tools:node="remove" />` , as that would presumably solve the issue with Mapbox. If it doesn't work, then we have to investigate further? @madhurgupta10

**Comment by madhurgupta10** [comment]
> ACCESS_BACKGROUND_LOCATION

I checked the merged manifest file and there is indeed no ACCESS_BACKGROUND_LOCATION permission requested. Not sure why we are getting this error

**Comment by nicolas-raoul** [comment]
[This developer had a similar problem](https://www.internetkatta.com/how-to-solve-location-policy-rejection-for-play-store), no ACCESS_BACKGROUND_LOCATION anywhere, and apparently this got their app approved:
![Screenshot from 2022-08-09 17-14-49](https://user-images.githubusercontent.com/99590/183599525-792a465d-3600-4918-907f-44d243c44c4c.png)

**Comment by misaochan** [comment]
@nicolas-raoul Do you think we could appeal, talk to Google and say that we checked our manifest and merged manifest, and there is no ACCESS_BACKGROUND_LOCATION in either? At the very least, they might be able to give us more details on what exactly makes them think we are accessing location data in the background. Otherwise it feels a bit like shooting in the dark, there are a lot of things we can try but it would be a long process to slowly go through them.

**Comment by madhurgupta10** [comment]
@misaochan Even in the permissions I don't see "Allow all time" request as stated in [docs](https://developer.android.com/training/location/permissions#request-background-location)

![image](https://user-images.githubusercontent.com/30932899/183834596-9c705d53-d6ac-43e0-9386-76fbccf07b87.png)

**Comment by misaochan** [comment]
Yes, indeed. :( I actually just tried promoting the beta to production anyway yesterday. It is still in review (more than 12h now), but we'll see what happens.

**Comment by whym** [comment]
[Playstore page](https://play.google.com/store/apps/details?id=fr.free.nrw.commons) has a link (under Developer contract) to https://github.com/commons-app/apps-android-commons/wiki/Privacy-policy which doesn't exist. I confirmed this with a real phone's Play Store app and desktop Firefox.

Is it supposed to be https://github.com/commons-app/commons-app-documentation/blob/master/android/Privacy-policy.md#privacy-policy ?

This might or might not be related to the particular rejection, but I think it's a good idea to fix it regardless. A faulty privacy policy link seems like a potential cause of rejection in other privacy related matters, too.

**Comment by nicolas-raoul** [comment]
@whym Fixed, thanks a lot! :-)

**Comment by misaochan** [comment]
App was rejected from prod about an hour ago (so after the privacy link was fixed).

**Comment by whym** [comment]
Is the link really fixed? Here is what I'm seeing. It could be my local cache or something, though.
![image](https://user-images.githubusercontent.com/10674/183901698-aa4cf135-b198-450d-8f52-7d947d478779.png)

**Comment by misaochan** [comment]
I believe in order to actually publish the new link, Google might have to approve our submission first, but I could be wrong.

**Comment by misaochan** [comment]
These are the results of my appeal to Google:

> Step 1: Review the policy violation with your app
> 
> We found that your app is not compliant with the Location Permissions policy, or we were unable to review and verify your in-app experience for compliance with this policy.
> 
> **Specifically, Prominent disclosure not found.**
> 
> Your app must display a prominent disclosure through a pop-up alert before your app’s location runtime permission. Based on our review, a prominent disclosure did not appear before the runtime permission. 
> 
> Please add a prominent disclosure before the runtime permission.
> 
> To meet the policy requirements, it is recommended that you reference one of the following example formats, the second example includes the use of location for ads (choose the most relevant phrasing):
> 
> “[This app] collects location data to enable ["feature"], ["feature"], & ["feature"] even when the app is closed or not in use.” 
> “[This app] collects location data to enable ["feature"], ["feature"], & ["feature"] even when the app is closed or not in use and it is also used to support advertising.”
> Example: “Fitness Funds collects location data to enable fitness tracking even when the app is closed or not in use.”
> 
> The language in the disclosure MUST include the following elements:
> 
> The term “location”
> Indication that the nature of usage is in the background by using one of the following phrases “background” / “when the app is closed” / “always in use” / “when the app is not in use”
> A list of all the features that use location in the background
> If you extend permitted usage to ads, you must include the following: “used to provide ads / support advertising / support ads.” (Choose the most accurate phrasing)
> Remember, your prominent disclosure must:
> 
> Appear before your app’s location runtime permission
> Include an example format listed above
> Include any other details necessary to make it clear to the user how and why you are using location in the background. While additional content is permitted, it should not cause the required content to not be immediately visible
> If portions of your app are restricted, please share a link to a video of your app that demonstrates the in-app feature’s functionality requesting location in the background and how a user would trigger the prominent disclosure and runtime permission (with user consent)
> Include buttons that are clear to the user if they are approving or denying access. If a user denies access and that would limit the app experience, then make sure to inform them on how to change preferences in the future
> Please ensure your prominent disclosure is included in the video required in the Developer console.
> 
> Learn more about prominent disclosure requirements in the [Developer Policy Center](https://support.google.com/googleplay/android-developer/answer/9799150?hl=en#zippy=%2Cstep-provide-prominent-in-app-disclosure) or in this [video](https://www.youtube.com/watch?v=b0I1Xq_iSK4).
> 
> If you have determined that your app does not require location in the background or if you believe your app no longer accesses location in the background, complete the following steps to ensure compliance: 
> 
> Ensure that the ACCESS_BACKGROUND_LOCATION permission has been removed from your app APK or app bundle. 
> If you are using ACCESS_COARSE_LOCATION or ACCESS_FINE_LOCATION, examine your code paths and ensure usage is restricted to foreground purposes only ([learn more](https://developer.android.com/training/location/background))
> If you previously had any non-compliant APKs accessing background location, make sure the non-compliant versions are not in any of your current releases, even if you do not use certain tracks. 
> Go to Release > App bundle explorer page to check whether a certain version is active
> When submitting a new APK to supersede the previous, non-compliant APK, please make sure the non-compliant app bundles or APKs are under the Not included section before rolling out the new release. 
> For further guidance, please see the "Not included (app bundles and APKs)" section in this [Play Console Help article](https://support.google.com/googleplay/android-developer/answer/9859348#zippy=%2Cnot-included-app-bundles-and-apks). 
> Please ensure that any new, compliant release is [rolled out to 100%](https://support.google.com/googleplay/android-developer/answer/6346149) and completely deactivates non-compliant APKs. 
> If you are still facing issues after examining your code paths and restricting usage to foreground purposes only, please review any third party SDKs used in the app that may be accessing location in the background.
> 
> For additional help, you can review the following resources: 
> 
> [Location permissions](https://support.google.com/googleplay/android-developer/answer/9888170#location-permissions) - Developer Policy Center
> [Requesting access to background location](https://support.google.com/googleplay/android-developer/answer/9799150) - Play Console Help
> [Evaluate background location access](https://developer.android.com/training/location/background#evaluate) & [Background location access checklist](https://developer.android.com/training/location/background#checklist) - Documentation for app developers
> [Google Play PolicyBytes on YouTube](https://www.youtube.com/watch?v=TDF-GvnKC18)
> Please update your app to fix this issue. You may also want to double check that your app complies with all other [Developer Program Policies](https://play.google.com/about/developer-content-policy/).

**Comment by madhurgupta10** [comment]
@misaochan one thing I understand is that we should add a dialog before the permission dialog which tells the user why we need this permission.

Apart from that, I don't really understand what's causing the background location :(

**Comment by misaochan** [comment]
@madhurgupta10 Yes, as I mentioned on Zulip, the email sounds like the prominent disclosure is their primary issue, and resolving that may allow our release to go through. We can leave this issue open for further investigation, should anyone wish to take it up in the future.

How about we add a dialog before the location permission that says: “The Commons app collects location data in the background to display nearby places that need photos"? This seems like it would satisfy their format.

**Comment by madhurgupta10** [comment]
This should be acceptable I guess. Also I think Google's definition for background location not necessarily mean that only apps that use location when they are closed are considered in this category but also like our app which uses location when app is not in foreground (like nearby walk) will also falls under this category so a more clear prompt message should be acceptable to Google.

**Comment by misaochan** [comment]
Ah yeah, makes sense. Do you think you could implement this dialog please?

**Comment by misaochan** [comment]
Still rejected. I appealed again...

> Feature does not meet requirements for background location access
> Based on our review, your declared feature does not meet the requirements for background location access.
> 
> Please remove the background location permission requested and submit an update to your app. When declaring a feature for background location access, please note the following:
> Your selected feature should deliver clear value to the user and be important to the core functionality, or main purpose of the app. Without this core feature, the app is “broken” or rendered unusable.
> You should also consider if users would expect the app to access their location in the background, and if you can deliver the same experience without accessing location in the background.

**Comment by misaochan** [comment]
The appeal has been declined. Oddly, v4.0.3 still remains up on beta, but v4.0.4 cannot be pushed to either beta or production.

> Status: Latest app update not available on Google Play
> 
> I’ve reviewed your appeal request and found that your app, Wikimedia Commons, (fr.free.nrw.commons), still violates Google Play Policy. If you submitted an update to an existing app, the version published prior to the update is still available on Google Play. I’ve included details below about the specific issue with your app and what you can do to get your app back on Google Play.
> 
> Step 1: Review the policy violation with your app
> 
> We found that your app is not compliant with the Location Permissions policy, or we were unable to review and verify your in-app experience for compliance with this policy.
> 
> Specifically, Feature does not meet requirements for background location access
> 
> We have re-reviewed your app, but your declared feature still does not meet the requirements to access location in the background because of the following reason:
> 
> It is possible to deliver a similar experience without access to location in the background
> Please note that we use multiple factors for evaluation, which results in a high bar for access to location in the background. You can refer to this [Help Center article](https://support.google.com/googleplay/android-developer/answer/9799150) for information on the requirements for accessing location in the background. 
> 
> Because of the restrictions on this policy, it is unlikely that your app will be approved for access to location in the background. Therefore, please remove the request to access location in the background and submit the update to your app.
> 
> If you are using ACCESS_COARSE_LOCATION or ACCESS_FINE_LOCATION, examine your code paths and restrict usage to foreground purposes only. Kindly check if you are using a 3rd party SDK that is using BG Location as well.
> (Learn more: https://developer.android.com/training/location/background)
> 
> Lastly, you should declare that you aren’t using background location by selecting “No” to the question “Does your app access location in the background in APKs or app bundles targeting Android 9 or older?"
> 
> For additional help, you can review the following resources: 
> 
> [Location permissions](https://support.google.com/googleplay/android-developer/answer/9888170#location-permissions) - Developer Policy Center
> [Requesting access to background location](https://support.google.com/googleplay/android-developer/answer/9799150) - Play Console Help
> [Evaluate background location access](https://developer.android.com/training/location/background#evaluate) & [Background location access checklist](https://developer.android.com/training/location/background#checklist) - Documentation for app developers
> [Google Play PolicyBytes on YouTube](https://www.youtube.com/watch?v=TDF-GvnKC18)
> Please update your app to fix this issue. You may also want to double check that your app complies with all other [Developer Program Policies](https://play.google.com/about/developer-content-policy/).
> 
> Step 2:  Submit a compliant update or remove permissions
> 
> Make the necessary updates to address the issue(s) identified above.
> To continue using location in the background, submit or update your [Developer Permission Declaration](https://play.google.com/console/developers/app/app-content/permission-declarations) along with a compliant version of your app in [Play Console](https://play.google.com/apps/publish/) for approval.
> If your app is not eligible to access location in the background or does not meet requirements for accessing location in the background, please remove the permission from your manifest and in-app functionality.
> Double check that your app is compliant with all other [Developer Program Policies](https://play.google.com/about/developer-content-policy/).
> Sign in to [Play Console](https://play.google.com/apps/publish/) and submit the update to your app.
> Please make sure to deactivate any non-compliant APKs. To do this, you can create a new release and upload a compliant APK to each track containing the non-compliant APKs. Be sure to increment the APK version code and set the release to 100% rollout.

**Comment by nicolas-raoul** [comment]
Apparently, because we define `minSdkVersion 19` (which is less than `29`), ACCESS_FINE_LOCATION [implicitly implies background location permission](https://stackoverflow.com/a/67392120/226958), so we must follow all the Google Play rules for that.

So it seems that we can either:
- In all places where location service is used, tell the user that we use background location.
- Or bump `minSdkVersion` to 29. That's Android 10 (released September 2019). That would mean 10541-2936-2886-2372=2347 of our users not being able to update (that's 2347/10541=22% of our users). Here "users" means people who have used the app in the last 30 days. Screenshot from the [Google Play Console](https://play.google.com/console/u/0/developers/6169333749249604352/app/4973046028707297472/statistics?metrics=ACTIVE_USERS-ALL-UNIQUE-PER_INTERVAL-DAY&dimension=OS_VERSION&dimensionValues=OVERALL%2C31%2C30%2C29&dateRange=2022_7_21-2022_8_19&tab=APP_STATISTICS&ctpMetric=DAU_MAU-ACQUISITION_UNSPECIFIED-COUNT_UNSPECIFIED-CALCULATION_UNSPECIFIED-DAY&ctpDateRange=2022_7_21-2022_8_19&ctpDimension=COUNTRY&ctpDimensionValue=OVERALL&ctpPeersetKey=3%3A3e613dcd79f81a2a):

<img width="283" alt="Screen Shot 2022-08-20 at 19 50 40" src="https://user-images.githubusercontent.com/99590/185742632-e304aa69-5328-415d-8424-d5779496dd33.png">

I have always been reluctant to increase minSdkVersion because our most valuable users (those in underrepresented countries/communities) are less likely to have recent phones. But if it is temporary in order to get important updates out to most users while adding the necessary permission popups, I am all for it :-)

**Comment by misaochan** [comment]
> In all places where location service is used, tell the user that we use background location.

Unfortunately I don't think this would solve the problem, as the reviewer explicitly said that we aren't eligible for background location at all because "It is possible to deliver a similar experience without access to location in the background".

> Or bump minSdkVersion to 29

Hmm, thanks, I was actually unaware that this might solve the issue. I agree, it would be extremely rough to drop 22% of our users, but this probably beats not being able to release updates at all. In a future grant, the team might consider requesting funds to look for solutions that allow min SDK to be reduced again.

I am willing to try one last update that includes bumping min SDK. @madhurgupta10 @neslihanturan ?

**Comment by madhurgupta10** [comment]
I suspect bumping up the SDK version will result in unexpected crashes whenever location service tries to get location in background.

**Comment by misaochan** [comment]
We don't actually access location in background, though? According to the SOF thread, it is just implied by our min SDK.

**Comment by nicolas-raoul** [comment]
> reviewer explicitly said [...]

Sorry I had missed that :-( Thanks for the hint!

> if someone's phone is <Android 10, can they not access the app at all, or can they still access the old version?

Existing users with an old Android will keep the out-of-date app. But new users with an old Android will not be able to find the app in Google Play.

**Comment by madhurgupta10** [comment]
> We don't actually access location in background, though? According to the SOF thread, it is just implied by our min SDK.

Google's definition of background location is quite confusing but we can try with this change

**Comment by misaochan** [comment]
I tested it and no additional crashes that I can see so far. Will try releasing with this.

**Comment by madhurgupta10** [comment]
@misaochan Also I think we should remove "in the background" from the permission prompt

**Comment by misaochan** [comment]
Good point!

**Comment by madhurgupta10** [comment]
@misaochan I think just updating the string of "location_permission_rationale_nearby" should be sufficient since my commit also fixes the bug where prompt was not shown before the permission request (Google's Policy)

**Comment by misaochan** [comment]
Seems like Nicolas's suggestion works, based on the email from google. I just need to "cleanse" all the other tracks and hopefully then it should be fine


Annoyingly, it does mean that the latest version has to go directly to prod without testing
. Hopefully there won't be too many issues with that

**Comment by misaochan** [comment]
It did not work, for some extremely perplexing reason. Even though version 1031 was not listed in the previous email from google re: non-compliance. I submitted an appeal, but don't really have much faith in it at this point.

Unfortunately that is it for us, we have put in a large number of unpaid hours to try and fix this and our final report is way overdue, so if the appeal does not succeed, I will document it and proceed with finalizing the grant anyway.

**Comment by misaochan** [comment]
Wow finally a useful appeal review! I think that did the trick, looks like v4.0.5 is out in production and beta!

**Comment by nicolas-raoul** [comment]
Great! And apologies if the faulty questionnaire was by me 😔
I think it would be a good idea to export your questionnaire answers and put the file in version control.

**Comment by misaochan** [comment]
Great idea! There are quite a lot of files though. Probably need a new folder for them, or just in the documentation repo?

**Comment by nicolas-raoul** [comment]
How about a new folder in
https://github.com/commons-app/marketing/tree/master/play-store-listing ?

**Comment by misaochan** [comment]
LGTM. :)

**Comment by otionce** [comment]
Hey guys, I was just looking at how to get involved with open source projects, and I stumbled on this one. 

I tried to catch up on what has been going on with the app. 
I downloaded it, and after being prompted to use the location services, I answer only while using this app, and then i got this weird bug where it would repeatedly ask permission over and over, and I couldn't even answer fast enough. 

I am new to the whole open source project thing, but just something that I noticed.
